pgmax/groups/enumeration.py

```diff
@@ -0,0 +1,459 @@
+
+        return flat_data
+
+    def unflatten(
+        self, flat_data: Union[np.ndarray, jnp.ndarray]
+    ) -> Union[np.ndarray, jnp.ndarray]:
+        """Function that recovers meaningful structured data from internal flat data array
+
+        Args:
+            flat_data: Internal flat data array.
+
+        Returns:
+            Meaningful structured data. Should be an array of shape (num_val_configs,)
+                (for shared log potentials) or (num_factors, num_val_configs) (for log potentials)
+                or (num_factors, num_edge_states) (for ftov messages).
+
+        Raises:
+            ValueError if:
+                (1) flat_data is not a 1D array
+                (2) flat_data is not of the right shape
+        """
+        if flat_data.ndim != 1:
+            raise ValueError(
+                f"Can only unflatten 1D array. Got a {flat_data.ndim}D array."
+            )
+
+        num_factors = len(self.factors)
+        if flat_data.size == num_factors * self.factor_configs.shape[0]:
+            data = flat_data.reshape(
+                (num_factors, self.factor_configs.shape[0]),
+            )
+        elif flat_data.size == num_factors * np.sum(self.factors[0].edges_num_states):
+            data = flat_data.reshape(
+                (num_factors, np.sum(self.factors[0].edges_num_states))
+            )
+        else:
+            raise ValueError(
+                f"flat_data should be compatible with shape {(num_factors, self.factor_configs.shape[0])} "
+                f"or {(num_factors, np.sum(self.factors[0].edges_num_states))}. Got {flat_data.shape}."
+            )
+
+        return data
+
+
+@dataclass(frozen=True, eq=False)
+class PairwiseFactorGroup(groups.FactorGroup):
+    """Class to represent a group of EnumerationFactors where each factor connects to
+    two different variables.
+
+    All factors in the group are assumed to be such that all possible configuration of the two
+    variable's states are valid. Additionally, all factors in the group are assumed to share
+    the same potential function and to be connected to variables from VariableGroups within
+    one CompositeVariableGroup.
+
+    Args:
+        log_potential_matrix: array of shape (var1.num_states, var2.num_states),
+            where var1 and var2 are the 2 VariableGroups (that may refer to the same
+            VariableGroup) whose names are present in each sub-list from self.variable_names_for_factors.
+        Raises:
+            ValueError if:
+                (1) The specified log_potential_matrix is not a 2D or 3D array.
+                (2) Some pairwise factors connect to less or more than 2 variables.
+                (3) The specified log_potential_matrix does not match the number of factors.
+                (4) The specified log_potential_matrix does not match the number of variable states of the
+                    variables in the factors.
+    """
+
+    factor_configs: np.array = field(init=False)
```

Comment:
If we get rid of `factor_configs` in the parent class we don't need to do this awkward thing here

Reply:
Same, it's mypy related